java-squid/src/main/java/org/sonar/java/xml/XmlCheckContextImpl.java

```diff
@@ -51,15 +56,57 @@ public File getFile() {
   }
 
   @Override
-  public NodeList evaluateXPathExpression(String expression) throws XPathExpressionException {
-    XPath xPath = XPathFactory.newInstance().newXPath();
-    return (NodeList) xPath.compile(expression).evaluate(document, XPathConstants.NODESET);
+  public XPathExpression compile(String expression) throws XPathExpressionException {
+    return xPath.compile(expression);
   }
 
   @Override
-  public NodeList evaluateXPathExpressionFromNode(Node node, String expression) throws XPathExpressionException {
-    XPath xPath = XPathFactory.newInstance().newXPath();
-    return (NodeList) xPath.compile(expression).evaluate(node, XPathConstants.NODESET);
+  public Iterable<Node> evaluateOnDocument(XPathExpression expression) throws XPathExpressionException {
+    return evaluate(expression, document);
+  }
+
+  @Override
+  public Iterable<Node> evaluate(XPathExpression expression, Node node) throws XPathExpressionException {
+    NodeList nodeList = (NodeList) expression.evaluate(node, XPathConstants.NODESET);
+    if (nodeList.getLength() == 0) {
+      return ImmutableList.of();
+    }
+    return new NodeListIterable(nodeList);
+  }
+
+  private static class NodeListIterable implements Iterable<Node> {
+
+    private final NodeList nodeList;
+    private final int length;
+
+    public NodeListIterable(NodeList nodeList) {
+      this.nodeList = nodeList;
+      this.length = nodeList.getLength();
+    }
+
+    @Override
+    public Iterator<Node> iterator() {
+      return new Iterator<Node>() {
+        private int count = 0;
+
+        @Override
+        public boolean hasNext() {
+          return count < length;
+        }
+
+        @Override
+        public Node next() {
+          if (count > length)
+            throw new NoSuchElementException();
+          return nodeList.item(count++);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract this increment or decrement operator into a dedicated statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS881)